concurrency/BaseFilterAdapter.java

```diff
@@ -0,0 +1,645 @@
+        public String directoryType;
+        public String displayName;
+        public String accountName;
+        public String accountType;
+        public boolean loading;
+        public CharSequence constraint;
+        public DirectoryPartitionFilter filter;
+
+        public DirectoryPartition() {
+            super(false, false);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == null) {
+                return false;
+            }
+            if (obj == this) {
+                return true;
+            }
+            if (obj instanceof DirectoryPartition) {
+                return directoryId == ((DirectoryPartition) obj).directoryId;
+            }
+            return false;
+        }
+
+        @Override
+        public int hashCode() {
+            return (int) directoryId;
+        }
+    }
+
+    private static class DirectoryListQuery {
+
+        // TODO: revert to references to the Directory class as soon as the
+        // issue with the dependency on SDK 8 is resolved
+        public static final Uri URI = Uri.withAppendedPath(ContactsContract.AUTHORITY_URI,
+            "directories");
+        private static final String DIRECTORY_ID = "_id";
+        private static final String DIRECTORY_ACCOUNT_NAME = "accountName";
+        private static final String DIRECTORY_ACCOUNT_TYPE = "accountType";
+        private static final String DIRECTORY_DISPLAY_NAME = "displayName";
+        private static final String DIRECTORY_PACKAGE_NAME = "packageName";
+        private static final String DIRECTORY_TYPE_RESOURCE_ID = "typeResourceId";
+
+        public static final String[] PROJECTION = { DIRECTORY_ID, // 0
+            DIRECTORY_ACCOUNT_NAME, // 1
+            DIRECTORY_ACCOUNT_TYPE, // 2
+            DIRECTORY_DISPLAY_NAME, // 3
+            DIRECTORY_PACKAGE_NAME, // 4
+            DIRECTORY_TYPE_RESOURCE_ID, // 5
+        };
+
+        public static final int ID = 0;
+        public static final int ACCOUNT_NAME = 1;
+        public static final int ACCOUNT_TYPE = 2;
+        public static final int DISPLAY_NAME = 3;
+        public static final int PACKAGE_NAME = 4;
+        public static final int TYPE_RESOURCE_ID = 5;
+    }
+
+    /**
+     * A fake column name that indicates a "Searching..." item in the list.
+     */
+    private static final String SEARCHING_CURSOR_MARKER = "searching";
+
+    /**
+     * An asynchronous filter used for loading two data sets: email rows from the local contact
+     * provider and the list of {@link Directory}'s.
+     */
+    private final class DefaultPartitionFilter extends Filter {
+
+        @Override
+        protected FilterResults performFiltering(CharSequence constraint) {
+            if (constraint == null) {
+                return null;
+            }
+            Cursor directoryCursor = null;
+            if (!mDirectoriesLoaded) {
+                directoryCursor = mContentResolver.query(DirectoryListQuery.URI,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *directoryCursor*. Currently, there are execution paths that do not contain closure statements. Either a) close *directoryCursor* in a try-finally block or b) close the resource by declaring *directoryCursor* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).